### PR TITLE
SPARK-11258 Converting a Spark DataFrame into an R data.frame is slow / requires a lot of memory

### PR DESCRIPTION
**Pull request by FRosner**
https://issues.apache.org/jira/browse/SPARK-11258

I was not able to locate an existing unit test for this function so I wrote one.

### CODE REVIEW

**Comment by sun-rui** [comment]
Instead of creating a new testsuite in Scala, you can add a new test case in R, using
callJStatic to invoke "dfToCols" on the Scala side.

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala [hunk 1]
```diff
@@ -130,16 +130,17 @@ private[r] object SQLUtils {
...
+    val numRows = localDF.length
 
-    // result is Array[Array[Any]]
-    (0 until numCols).map { colIdx =>
-      localDF.map { row =>
-        row(colIdx)
-      }
-    }.toArray
+    val colArray = new Array[Array[Any]](numCols)
+    for (colNo <- 0 to (numCols - 1)) colArray.update(colNo, new Array[Any](numRows))
```
Comment (sun-rui) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: What about the following code:
    for (colIdx <- 0 until numCols) {
      colArray(colIdx) = new Array[Any](numRows)
    }
    for (rowIdx <- 0 until numRows; colIdx <- 0 until numCols) {
      colArray(colIdx)(rowIdx) = localDF(rowIdx)(colIdx)
    }

**Comment by FRosner** [comment]
Thanks for the reply.

> Instead of creating a new testsuite in Scala, you can add a new test case in R, using callJStatic to invoke "dfToCols" on the Scala side.

What would be the advantage of this? I think it is good practice to have unit tests for all public methods. Using `callJStatic` seems like tackling this in a roundabout way.

Where can I find the test cases in R?

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala (hunk 1 above)
Comment (FRosner) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: Seems also plausible. Do you want me to change it?

**Comment by sun-rui** [comment]
dfToCols is meant to be called from R side, it makes sense to test it in R. Having a test case  in R for it can test not only the logic in Scala, but also helps to test other R Scala brige logic including serde. There is no test case in Scala for SparkR for now, of course, we will add if really needed

test cases in R are at R/pkg/inst/tests

**Comment by felixcheung** [comment]
Do you have benchmark numbers for this change?

**Comment by FRosner** [comment]
@felixcheung the ticket is mentioning two numbers (https://issues.apache.org/jira/browse/SPARK-11258#) in the description. For a small data frame it is already a speedup of 4. But anyway - quadratic complexity is the problem as you run into issues with increasing number of columns quickly. We were not even able to load one of our data frames (> 1 Million rows, several hundred of columns) with the old method but it ran through with the new one.

I can provide some more benchmarks later today.

**Comment by FRosner** [comment]
@sun-rui so there is an existing R test case for this method? In the end I just did a refactoring so existing tests should not break.

The idea of unit tests however is to test one thing. A unit test for a method should test exactly what this method is doing and not other related things like serde. If serde breaks, you still want to test for this method to succeed, because it is working. And since it is public API, anyone using Spark can use it.

**Comment by FRosner** [comment]
@felixcheung it is strange. When I did some performance checks today, the run time did not seem to be quadradic. Is it maybe because the map transformation in the original code is lazy and only executed when you actually access the data?

I can tell you that the original implementation did not succeed on our file while mine did. I will investigate a bit more. Maybe someone else has an opinion?

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala [hunk 2]
```diff
@@ -130,16 +130,18 @@ private[r] object SQLUtils {
...
+    val localDF: Array[Row] = df.collect()
     val numCols = df.columns.length
+    val numRows = localDF.length
 
-    // result is Array[Array[Any]]
-    (0 until numCols).map { colIdx =>
-      localDF.map { row =>
-        row(colIdx)
+    val colArray = new Array[Array[Any]](numCols)
+    for (colNo <- 0 until numCols) {
```
Comment (shivaram) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: Using a while loop here instead of a for loop should also help with performance

Comment (FRosner) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: Yeah we might give this a try but I don't think that the loop itself is the problem but rather the stuff that is going on in map and then .toArray. I will investigate a bit more over the week end.

**Comment by Gerrrr** [comment]
The problem might be related to GC as a current implementation of `dfToCols` allocates memory for the map result and the array while @FRosner implementation only allocates memory for the array.

**Comment by shivaram** [comment]
Jenkins, ok to test

**Comment by sun-rui** [comment]
@FRosner, there is no test case dedicated for dfToCols in R now. I am still reluctant to add a new test case in Scala simply for dfToCols. I'd like to leave the decision to other's comments.

**Comment by sun-rui** [comment]
@FRosner, it seems that you are investigating why the performance of the old implementation does not behave as you expect? No matter the investigation result, your new version does have better performance. If we can't find a better one than you, let's have it merged first:)

**Comment by FRosner** [comment]
@sun-rui thanks for your comments. You should not hesitate to add unit tests for functions that are not tested, yet. Tests help other developers (like me) to change stuff in the code without the immediate fear of unforseen consequences. They also document the API. As this function is public, it should have a test.

I will look into the build failure reason now @shivaram. I was not able to run all tests locally as I got out of memory problems, although I changed the maven settings as indicated in the documentation.

**Comment by FRosner** [comment]
@shivaram I added the missing license header. Sorry for missing this one.

**Comment by sun-rui** [comment]
@FRosner, dfToCols is not a public API, and is now only a helper function for SparkR. Could you add a private modifier for it?

**Comment by FRosner** [comment]
@sun-rui I am not sure how callJStatic works but I can imagine that making a method private does not allow it to be called - neither from the test, nor from R. Why would you not want to have a unit test. Test driven development is a good invention and helps developers (especially in OSS) to trust the code and make changes much easier.

**Comment by sun-rui** [comment]
@FRosner, I am not denying a unit test:). My point is that it seems not necessary to add a Scala unit case for dfToCols, which is dedicated for SparkR now and its logic is simple. Adding a R test case is fine for me, but it is a plus but not compelling, as its functionality is covered by existing test cases for collect().

As for private modifier, I could be wrong, forget it:)

**Comment by shivaram** [comment]
The class is already `private[r]` so it should be fine.

@FRosner Is the change good to go or are you doing any more perf tests ?

**Comment by felixcheung** [comment]
@FRosner I assume this is faster than the original implementation?

**Comment by FRosner** [comment]
In my opinion we are good to go if there are no objections from your side. 
-- 
Sent from my phone. Please excuse my brevity.

Shivaram Venkataraman <notifications@github.com> wrote:

The class is already private[r] so it should be fine.

@FRosner Is the change good to go or are you doing any more perf tests ?

**Comment by FRosner** [comment]
@felixcheung the runtime performance tests I conducted came to the result that the performance gain is minimal if there is enough RAM available. It was faster all the time but only a few per cent. However when we tried to load a huge parquet file (70 GB uncompressed), we ran into problems with the old implementation. As @Gerrrr indicated it might be related to the map and then .toArray, which causes memory overhead. I think when it comes to collecting data to the driver we should ensure that we have a small memory consumption.

**Comment by shivaram** [comment]
Ok I see -- we used to make a Seq[Array[Any]] and then call toArray on it before and now we just have Array[Array[Any]]. Alright change LGTM. Merging this

**Pull request closed without merging**